File: 2170/2170.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

int minimumOperations(int* nums, int numsSize){
    if (numsSize < 2) {
        return 0;
    }
    // count odd index and even index
    int most1[2] = {0,0}, most2[2] = {0,0};
    int count[2][100001]; // count[0][x] for even, count[1][x] for odd
    memset(count, 0, sizeof(count));
    int idx;
    for (idx = 0; idx < numsSize; ++idx) {
        int n = nums[idx];
        int eo = idx & 1;
        ++count[eo][n];
        if (n == most1[eo]) {
            continue;
        } else if (count[eo][n] > count[eo][most1[eo]]) {
            most2[eo] = most1[eo];
            most1[eo] = n;
        } else if (count[eo][n] > count[eo][most2[eo]]) {
            most2[eo] = n;
        }
        /* printf("%d: %d %d\n", n, counto[n], counte[n]); */
        /* printf("o: %d %d, e: %d %d\n", odd1, odd2, even1, even2); */
    }
    if (most1[0] == most1[1]) {
        if (count[0][most1[0]] + count[1][most2[1]] > count[0][most2[0]] + count[1][most1[1]]) {
            return numsSize - count[0][most1[0]] - count[1][most2[1]];
        } else {
            return numsSize - count[0][most2[0]] - count[1][most1[1]];
        }
    }
    return numsSize - count[0][most1[0]] - count[1][most1[1]];
}

int main() {
    int nums[] = {3,3,3,3,3,3,3};
    printf("%d", minimumOperations(nums,7));
}

File: 2170/2170.rs
impl Solution {
    pub fn minimum_operations(nums: Vec<i32>) -> i32 {
        // freq[is_odd][first or second] used to record first or second frequence number
        // freq[0][0] for first freq even
        // freq[0][1] for second freq even
        // freq[1][0] for first freq odd
        // freq[1][1] for second freq odd
        let mut freq: [[i32; 2]; 2] = [[0; 2]; 2];
        // count[is_odd][n], count the frequence of odd or even n
        let mut count: [[i32; 100001]; 2] = [[0; 100001]; 2];
        for (idx, &n) in nums.iter().enumerate() {
            let is_odd: usize = (idx & 1) as usize;
            count[is_odd][n as usize] += 1;
            if count[is_odd][n as usize] >= count[is_odd][freq[is_odd][0] as usize] {
                // Check if freq[x][0] should shift to freq[x][1] or not
                if freq[is_odd][0] != n {
                    freq[is_odd][1] = freq[is_odd][0];
                }
                freq[is_odd][0] = n;
            } else if count[is_odd][n as usize] > count[is_odd][freq[is_odd][1] as usize] {
                freq[is_odd][1] = n;
            }
        }
        if freq[0][0] == freq[1][0] {
            if count[0][freq[0][1] as usize] + count[1][freq[1][0] as usize] >
                count[0][freq[0][0] as usize] + count[1][freq[1][1] as usize] {
                    return nums.len() as i32 - count[0][freq[0][1] as usize] - count[1][freq[1][0] as usize];
                }
            return nums.len() as i32 - count[0][freq[0][0] as usize] - count[1][freq[1][1] as usize]
        }
        nums.len() as i32 - count[0][freq[0][0] as usize] - count[1][freq[1][0] as usize]
    }
}
